Source/WebKit/UIProcess/BrowsingContextGroup.cpp

```diff
@@ -68,17 +77,20 @@ void BrowsingContextGroup::sharedProcessForSite(WebsiteDataStore& websiteDataSto
         if (RefPtr frameProcess = protectedThis->m_sharedProcess.get())
             return completionHandler(frameProcess.get());
 
-        Ref process = pageConfiguration->protectedProcessPool()->processForSite(websiteDataStore.get(), site, lockdownMode, enhancedSecurity, pageConfiguration.get(), ProcessSwapDisposition::Other);
-        Ref frameProcess = FrameProcess::create(process, protectedThis, std::nullopt, preferences, InjectBrowsingContextIntoProcess::Yes);
+        Ref process = pageConfiguration->protectedProcessPool()->processForSite(websiteDataStore.get(), WebProcessPool::IsSharedProcess::Yes, site, mainFrameSite, domainsWithUserInteraction, lockdownMode, enhancedSecurity, pageConfiguration.get(), ProcessSwapDisposition::Other);
+        Ref frameProcess = FrameProcess::create(process, protectedThis, std::nullopt, mainFrameSite, preferences, InjectBrowsingContextIntoProcess::Yes);
+        ASSERT(frameProcess->isSharedProcess());
+        ASSERT(frameProcess->process().isSharedProcess());
+        frameProcess->process().addSharedProcessDomain(site.domain());
         protectedThis->m_sharedProcess = frameProcess.ptr();
         completionHandler(frameProcess.ptr());
     });
 }
 
-Ref<FrameProcess> BrowsingContextGroup::ensureProcessForSite(const Site& site, WebProcessProxy& process, const WebPreferences& preferences, InjectBrowsingContextIntoProcess injectBrowsingContextIntoProcess)
+Ref<FrameProcess> BrowsingContextGroup::ensureProcessForSite(const Site& site, const Site& mainFrameSite, WebProcessProxy& process, const WebPreferences& preferences, InjectBrowsingContextIntoProcess injectBrowsingContextIntoProcess)
```

Comment:
When are you supposed to use `ensureProcessForSite` instead of `sharedProcessForSite`? I tried reading through the callsites and don't understand the logic.